### DISCUSSION TITLE
Identify token/Request in handleData?

### DISCUSSION THREAD

**Discussion by schuppeste** [eModbus/eModbus#365] [General]
i read 5 Registers by addRequest, how i differentiate which register is handled?

**Comment by Miq1** [comment]
If you will read 5 registers in one request, you will have the register values in the response in a row, lowest register first. The get the respective values you will want to use the ``get()`` method with increasing offsets (+2 for each register value in sequence) to get these one by one. ``get()`` will return the offset incremented according the data type, so better use that. There are several applications of this technique in the examples, like
```
uint16_t offset = 2;

uint16_t reg1, reg2, reg3;
offset = response.get(offset, reg1);
offset = response.get(offset, reg2);
offset = response.get(offset, reg3);
```

**Reply by schuppeste** [reply]
Thanks for fast Reply.

I'm using ModbusClientTCP with Single Register Call.

My Workaround, i set the First Token Byte to Type/Identify.

AddRequest:
uint32_t token = millis() | (ident << 24);

handleData:
uint8_t ident = (token >> 24) & 0xff;

**Reply by Miq1** [reply]
Will not work... ``millis()`` is a 32 bit value, so even the first byte may change eventually. 

Better:
```
uint8_t ident = 84;
uint32_t token = (millis() << 8) | ident;
...
and
...
uint8_t ident = token & 0xFF;
```

**Reply by Miq1** [reply]
I take it from your post the registers you are trying to read are not consecutive or close at least?

I am asking because a single multi-register request will have much less overhead as compared to several (n times the TCP session open/close plus your ``ident`` handling) even if you will read a number of unnecessary registers in between.

**Reply by schuppeste** [reply]
Its okay for me, i Read 6 Registers every 30 Seconds.

It works. You can create your own Token, Random number or byte manipulated numbers. So i use the First byte to identify the function that adds the Requests.

**Reply by Miq1** [reply]
Whatever you like, but ``millis() | (ident << 24)`` will OR the ``ident`` value with whatever the first byte of ``millis()`` is - so the inverse will never be ``ident`` again...

**Reply by schuppeste** [reply]
You are Right. 

Its only a good Result if all Bits in highest Byte are Zero? ~17000 Seconds.

Maybe this: token=(ident <<24)  | (random)&0x00FFFFFF ;

I can use a Random Number or a simple Request Counter starts on 0 .. so i have 17Million Requests free in 3 Bytes.

Thanks for Reply.
